pkg/operator/health/checking_sync_wrapper.go

```diff
@@ -0,0 +1,39 @@
+package health
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+)
+
+// CheckingSyncWrapper wraps calls to the factory.SyncFunc in order to track when it last ran successfully.
+type CheckingSyncWrapper struct {
+	syncFunc          factory.SyncFunc
+	livenessThreshold time.Duration
+
+	// lastSuccessfulRun is updated with atomics, as updates can come from different goroutines
+	lastSuccessfulRun int64
```

Comment:
I'm probably missing something but I'm a little unclear on the need for atomics here.

Since each controller is constructing its own `CheckingSyncWrapper` struct that means the `Sync()` for each controller's goroutine would only update its own `checkingSyncWrapper.lastSuccessfulRun`.
So the controller goroutines aren't really sharing this state or variable right?

Reply:
Oh you probably meant the parallelism from the workers of each controller.
https://github.com/openshift/cluster-etcd-operator/blob/6ecd6d294d9b1bc713e674611e64fdea48202c44/pkg/operator/starter.go#L398-L416

Seems like we're not using any parallelism in any controller (nor would some of them tolerate that I imagine) but if that is ever the case in the future then the use of atomics makes more sense.

Reply:
yes and no, the value is read from the http handler called from the health probe. So that's where your concurrent access comes from. That's just to prevent stale cached reads.